### PR TITLE
feat: Support pending Pacts

### LINKED ISSUE #164: Support pending pacts
Hi,
Please add support for the new "pending pacts" feature:
https://pact.canny.io/feature-requests/p/pending-pacts-prevent-consumers-being-able-to-break-provider-builds-with-changed

I might be able to open a PR for this myself if I get some guidance.

### PR DESCRIPTION
**Pull request by m-aciek**
Closes #164.

### CODE REVIEW

**Comment by matthewbalvanz-wf** [comment]
@m-aciek `mock` should be part of the Python standard library from 3.3 and above, maybe we could remove that entirely now that support for Python 2.7 has been removed?
https://docs.python.org/3/library/unittest.mock.html

**Comment by m-aciek** [comment]
@matthewbalvanz-wf Unfortunately in 3.4 and 3.5 mock that is shipped with unittest has different API than in 3.8 so the asserts in VerifyWrapper fail.

File: pact/verify_wrapper.py [hunk 1]
```diff
@@ -169,6 +169,7 @@ def call_verify(self, *pacts, provider_base_url, provider, **kwargs):
 
         if(kwargs.get('verbose', False) is True):
             command.extend(['--verbose'])
+        command.extend(self._get_enable_pending_flag(kwargs.get('enable_pending')))
```
Comment (przemyslaw-jakielaszek) [inline pact/verify_wrapper.py:172]: _get_boolean_flag('pending')

File: pact/verify_wrapper.py [hunk 2]
```diff
@@ -169,6 +169,7 @@ def call_verify(self, *pacts, provider_base_url, provider, **kwargs):
 
         if(kwargs.get('verbose', False) is True):
```
Comment (przemyslaw-jakielaszek) [inline pact/verify_wrapper.py:170]: command.extend(self. _get_boolean_flag('verbose'))

**Review by przemyslaw-jakielaszek — CHANGES_REQUESTED** [review]

File: pact/cli/verify.py [hunk 3]
```diff
@@ -104,20 +104,23 @@
...
-    default=False,
-    help='Toggle verbose logging, defaults to False.')
 @click.option(
     'log_dir', '--log-dir',
     help='The directory for the pact.log file.')
 @click.option(
     'log_level', '--log-level',
     help='The logging level.')
+@click.option(
+    'enable_pending', '--enable-pending/--no-enable-pending',
```
[the remark this answers is not in the document]
Comment (m-aciek) [reply pact/cli/verify.py:114]: I wanted to replicate the flag that is used in underlying CLI and to be in line with flag name used in documentation https://docs.pact.io/pact_broker/advanced_topics/pending_pacts/#to-start-using-the-pending-pacts-feature.

I used default=None also to replicate usage of underlying CLI so to map False -> --enable…, True -> --no-enable…, None -> no flag.

File: pact/verify_wrapper.py [hunk 4]
```diff
@@ -199,3 +200,11 @@ def publish_results(self, provider_app_version, command):
         command.extend(["--provider-app-version",
                        provider_app_version,
                        "--publish-verification-results"])
+
+    @classmethod
+    def _get_enable_pending_flag(cls, flag_value):
```
[the remark this answers is not in the document]
Comment (m-aciek) [reply pact/verify_wrapper.py:205]: There is no such a flag as --no-verbose, so I think it is a premature optimization.

File: pact/verify_wrapper.py (hunk 1 above)
Comment (m-aciek) [reply pact/verify_wrapper.py:172]: >There is no such a flag as --no-verbose, and for just one case I think it is a premature optimisation.

File: pact/verify_wrapper.py (hunk 2 above)
Comment (m-aciek) [reply pact/verify_wrapper.py:170]: >There is no such a flag as --no-verbose, and for just one case I think it is a premature optimisation.

**Comment by m-aciek** [comment]
@matthewbalvanz-wf Summary of failing tests because of mock in 3.5 compatibility issues:

```
Error
Traceback (most recent call last):
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 605, in run
    testMethod()
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 1159, in patched
    return func(*args, **keywargs)
  File "/Users/maciek/projects/pact-python/tests/cli/test_verify.py", line 122, in test_broker_url_and_provider_required
    mock_wrapper.assert_called()
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 585, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called

Error
Traceback (most recent call last):
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 605, in run
    testMethod()
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 1159, in patched
    return func(*args, **keywargs)
  File "/Users/maciek/projects/pact-python/tests/cli/test_verify.py", line 339, in test_expand_directories_called
    mock_expand_dirs.assert_called_once()
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 585, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called_once

Error
Traceback (most recent call last):
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 605, in run
    testMethod()
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 1159, in patched
    return func(*args, **keywargs)
  File "/Users/maciek/projects/pact-python/tests/test_verifier.py", line 74, in test_expand_directories_called_for_pacts
    mock_expand_dir.assert_called_once()
  File "/Users/maciek/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 585, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called_once

SubTest failure: Traceback (most recent call last):
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 523, in subTest
    yield
  File "/…/projects/pact-python/tests/test_verifier.py", line 84, in test_passes_enable_pending_flag_value
    mock_wrapper.call_args.kwargs,
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 682, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : kwargs

Error
Traceback (most recent call last):
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/case.py", line 605, in run
    testMethod()
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 1159, in patched
    return func(*args, **keywargs)
  File "/…/pact-python/tests/test_verify_wrapper.py", line 160, in test_rerun_command_called
    mock_rerun_cmd.assert_called_once()
  File "/…/.pyenv/versions/3.5.9/lib/python3.5/unittest/mock.py", line 585, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called_once
```

**Comment by matthewbalvanz-wf** [comment]
Looks like `assert_called` and `assert_called_once` were added in 3.6, so it seems like using the library until that is end of life is the path forward.

File: pact/verify_wrapper.py (hunk 1 above)
Comment (przemyslaw-jakielaszek) [reply pact/verify_wrapper.py:172]: Maybe I am wrong, but I think there is a `no-verbose` flag.
https://github.com/pact-foundation/pact-python/pull/171/files#diff-99dbf564da4f139a45a61bf9c28b4f28R104

File: pact/cli/verify.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (przemyslaw-jakielaszek) [reply pact/cli/verify.py:114]: 1. you are right. It is said to use `enable-pending` convention: https://github.com/pact-foundation/pact_broker/issues/320
2. it is also said that flag should be set to False as a default value. Moreover I do not see a case when `no flag` differs from `no-enable-pending`

**Review by przemyslaw-jakielaszek — CHANGES_REQUESTED** [review]

File: pact/cli/verify.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-aciek) [reply pact/cli/verify.py:114]: Ok, so to be in line with the guidelines you linked and to make the implementation minimal, I will stick only with positive flag, set False as default and do not add --no-enable-pending to underlying library.

File: pact/verify_wrapper.py (hunk 1 above)
Comment (m-aciek) [reply pact/verify_wrapper.py:172]: It is lacking in the ruby verifier help message (see https://github.com/pact-foundation/pact-provider-verifier/blob/master/README.md), but I checked and indeed this ruby binary accepts --no-verbose. Although according to https://github.com/pact-foundation/pact-python/pull/171#discussion_r483937714 I'll drop the negative flag passing.

**Comment by m-aciek** [comment]
I've added Python 3.4 syntax compatibility and simplified code. Ready for review again.

**Comment by m-aciek** [comment]
I've fixed code issue with Click option.

**Comment by bethesque** [comment]
Would you mind leaving it as --enable-pending and --no-enable-pending in the python interface? When trying to write documentation that is applicable for all languages, it makes it very difficult when each of the 11 languages chooses a slightly different option name!

**Comment by elliottmurray** [comment]
@bethesque isn't the 'no' option redundant? Just wondering if it should be default off and you just ask --enable-pending but not sure what the plans are for this ongoing?

**Comment by m-aciek** [comment]
@bethesque @elliottmurray I had the same doubt and simplified implementation to avoid redundancy.

The use case for --no-enable-pending that I can think of would be having some default configuration that could be overridden in a command call.

I plan to implement support for --no-enable-pending on another branch and open new alternative pull request to avoid going round in circles in case of further discussion.

**Comment by m-aciek** [comment]
@bethesque @elliottmurray 

> The use case for --no-enable-pending that I can think of would be having some default configuration (e.g. in pyproject.toml) that could be overridden in a command call.

In this use case default for enable-pending flag would be `None`, meaning „not overriding some default configuration” in case of no of those flags are added. I'm assuming this in the implementation.

**Comment by elliottmurray** [comment]
I'm going to merge this and do a release later. It doesn't harm as far as I can tell and we can add the extra flag in another PR

**Comment by bethesque** [comment]
Sorry, I don't care about --no-enable-pending, just the --enable-pending (rather than --pending or whatever it was)

**Pull request merged**